packages/expo-clipboard/android/src/main/java/expo/modules/clipboard/ClipboardModule.kt

```diff
@@ -5,36 +5,91 @@ import android.content.ClipData
 import android.content.ClipboardManager
 import android.util.Log
 import androidx.core.os.bundleOf
+import expo.modules.core.errors.ModuleDestroyedException
 import expo.modules.core.utilities.ifNull
+import expo.modules.kotlin.Promise
 import expo.modules.kotlin.exception.CodedException
 
 import expo.modules.kotlin.modules.Module
 import expo.modules.kotlin.modules.ModuleDefinition
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import java.io.File
 
 private const val moduleName = "ExpoClipboard"
-const val clipboardChangedEventName = "onClipboardChanged"
 private val TAG = ClipboardModule::class.java.simpleName
 
-const val ERR_CLIPBOARD_UNAVAILABLE = "ERR_CLIPBOARD_UNAVAILABLE"
-
-class ClipboardUnavailableException :
-  CodedException(ERR_CLIPBOARD_UNAVAILABLE, "'CLIPBOARD_SERVICE' is unavailable on this device", null)
+// this must match the one from `res/xml/clipboard_provider_paths.xml`
+const val clipboardDirectoryName = ".clipboard"
+const val clipboardChangedEventName = "onClipboardChanged"
 
 class ClipboardModule : Module() {
   override fun definition() = ModuleDefinition {
     name(moduleName)
 
+    // region Strings
     function("getStringAsync") {
-      val clip = clipboardManager.primaryClip?.takeIf { it.itemCount >= 1 }
-      clip?.getItemAt(0)?.text ?: ""
+      clipboardManager.firstItem?.text ?: ""
     }
 
     function("setStringAsync") { content: String ->
       val clip = ClipData.newPlainText(null, content)
       clipboardManager.setPrimaryClip(clip)
       return@function true
     }
+    // endregion
+
+    // region Images
+    function("getImageAsync") { options: GetImageOptions, promise: Promise ->
+      val imageUri = clipboardManager
+        .takeIf { clipboardHasItemWithType("image/*") }
+        ?.firstItem
+        ?.uri
+        ?: return@function null
+
+      val exceptionHandler = CoroutineExceptionHandler { _, err ->
```

Comment:
Were there any specific reasons to use `CoroutineExceptionHandler` instead of `try-catch block`?

Reply:
Maybe in clipboard there are no such situations (however not sure about `runInterruptible`), but I encountered this in expo-image:
When having a single coroutine, a `try-catch` _inside_ it is sufficient. But when you have multiple nested coroutine launches, the outer try-catch won't catch the inner coroutine:
```kt
// launch outer coroutine
coroutineScope.launch {
  try {
    // launch the inner one
    // the situation would be a bit different with async { ... } but it's a different matter
    launch {
      throw RuntimeException("This won't be caught, its another coroutine")
    }
    throw RuntimeException("But this would be")
  } catch (...) { ... }
}
```
The `CoroutineExceptionHandler` however is able to catch all uncaught exceptions from nested coroutines - so it's good to have it at the function level to catch all uncaught `IOExceptions` somewhere in the coroutine logic.

Also, the `SupervisorJob` in the `moduleCoroutineScope` initializer prevents failing other async module functions (coroutines) when one of them fail (if they're running simultaneously, e.g. by running `Promise.all`). 

So I consider this as a good practice - if we ever have built-in coroutine support for `asyncFunction()` in sweet API, this is IMO a must-have.